job-server/src/spark.jobserver/WebApi.scala

```diff
@@ -253,25 +254,35 @@ class WebApi(system: ActorSystem,
           .map { contexts =>
             ctx.complete(removeProxyUserPrefix(authInfo, contexts)) }
       } ~
-        post {
-          /**
-           *  POST /contexts/<contextName>?<optional params> -
-           *    Creates a long-running context with contextName and options for context creation
-           *    All options are merged into the defaults in spark.context-settings
-           *
-           * @optional @param num-cpu-cores Int - Number of cores the context will use
-           * @optional @param memory-per-node String - -Xmx style string (512m, 1g, etc)
-           * for max memory per node
-           * @return the string "OK", or error if context exists or could not be initialized
-           */
+      post {
+        /**
+         *  POST /contexts/<contextName>?<optional params> -
+         *    Creates a long-running context with contextName and options for context creation
+         *    All options are merged into the defaults in spark.context-settings
+         *
+         * @optional @entity The POST entity should be a Typesafe Config format file with a
+         *            "spark.context-settings" block containing spark configs for the context.
+         * @optional @param num-cpu-cores Int - Number of cores the context will use
+         * @optional @param memory-per-node String - -Xmx style string (512m, 1g, etc)
+         * for max memory per node
+         * @return the string "OK", or error if context exists or could not be initialized
+         */
+        entity(as[String]) { configString =>
           path(Segment) { (contextName) =>
             // Enforce user context name to start with letters
             if (!contextName.head.isLetter) {
               complete(StatusCodes.BadRequest, errMap("context name must start with letters"))
             } else {
               parameterMap { (params) =>
-                val (cName, config) = determineProxyUser(ConfigFactory.parseMap(params.asJava).resolve(),
-                                                            authInfo, contextName)
+                // parse the config from the body, using url params as fallback values
+                val contextConfig = Try {
+                  ConfigFactory.parseString(configString).getConfig("spark.context-settings")
+                } getOrElse {
+                  ConfigFactory.empty
+                } withFallback {
+                  ConfigFactory.parseMap(params.asJava)
+                }
+                val (cName, config) = determineProxyUser(contextConfig.resolve(), authInfo, contextName)
```

Comment:
> I think it's a tiny bit better to have resolve() be called and assigned to contextConfig, instead of here, in case later someone tries to use it and forgets.

Good call, fixed!